File: Emulator/src/main/core/net/NioReactor.java
package core.net;

import core.net.amsc.MSEventHandler;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.channels.SelectionKey;
import java.nio.channels.Selector;
import java.nio.channels.ServerSocketChannel;
import java.nio.channels.SocketChannel;
import java.util.Iterator;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Handles (NIO-based) networking events using the reactor pattern.
 *
 * @author Emperor
 */
public final class NioReactor implements Runnable {

    /**
     * The executor service.
     */
    private final ExecutorService service;

    /**
     * The socket channel.
     */
    private ServerSocketConnection channel;

    /**
     * The I/O event handling instance.
     */
    private IoEventHandler eventHandler;

    /**
     * If the reactor is running.
     */
    private boolean running;

    private NioReactor(IoEventHandler eventHandler) {
        this.service = Executors.newSingleThreadScheduledExecutor();
        this.eventHandler = eventHandler;
    }

    /**
     * Configure nio reactor.
     *
     * @param port the port
     * @return the nio reactor
     * @throws IOException the io exception
     */
    public static NioReactor configure(int port) throws IOException {
        return configure(port, 1);
    }

    /**
     * Configure nio reactor.
     *
     * @param port     the port
     * @param poolSize the pool size
     * @return the nio reactor
     * @throws IOException the io exception
     */
    public static NioReactor configure(int port, int poolSize) throws IOException {
        NioReactor reactor = new NioReactor(new IoEventHandler(Executors.newFixedThreadPool(poolSize)));
        ServerSocketChannel channel = ServerSocketChannel.open();
        Selector selector = Selector.open();
        channel.bind(new InetSocketAddress(port));
        channel.configureBlocking(false);
        channel.register(selector, SelectionKey.OP_ACCEPT);
        reactor.channel = new ServerSocketConnection(selector, channel);
        return reactor;
    }

    /**
     * Connect nio reactor.
     *
     * @param address the address
     * @param port    the port
     * @return the nio reactor
     * @throws IOException the io exception
     */
    public static NioReactor connect(String address, int port) throws IOException {
        NioReactor reactor = new NioReactor(new MSEventHandler());
        Selector selector = Selector.open();
        SocketChannel channel = SocketChannel.open();
        channel.configureBlocking(false);
        channel.socket().setKeepAlive(true);
        channel.socket().setTcpNoDelay(true);
        channel.connect(new InetSocketAddress(address, port));
        channel.register(selector, SelectionKey.OP_CONNECT);
        reactor.channel = new ServerSocketConnection(selector, channel);
        return reactor;
    }

    /**
     * Start.
     */
    public void start() {
        running = true;
        service.execute(this);
    }

    @Override
    public void run() {
        Thread.currentThread().setName("NioReactor");
        while (running) {
            try {
                channel.getSelector().select();
            } catch (IOException e) {
                e.printStackTrace();
            }
            Iterator<SelectionKey> iterator = channel.getSelector().selectedKeys().iterator();
            while (iterator.hasNext()) {
                SelectionKey key = iterator.next();
                iterator.remove();
                try {
                    if (!key.isValid() || !key.channel().isOpen()) {
                        key.cancel();
                        continue;
                    }
                    if (key.isConnectable()) {
                        eventHandler.connect(key);
                    }
                    if (key.isAcceptable()) {
                        eventHandler.accept(key, channel.getSelector());
                    }
                    if (key.isReadable()) {
                        eventHandler.read(key);
                    } else if (key.isWritable()) {
                        eventHandler.write(key);
                    }
                } catch (Throwable t) {
                    eventHandler.disconnect(key, t);
                }
            }
        }
    }

    /**
     * Terminate.
     */
    public void terminate() {
        running = false;
    }

}

File: server/src/main/core/net/NioReactor.kt
package core.net

import core.net.amsc.MSEventHandler
import java.io.IOException
import java.net.InetSocketAddress
import java.nio.channels.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * A simple NIO-based reactor that handles non-blocking I/O operations
 * for both server and client socket channels.
 *
 * Used to accept, connect, read, and write events via a selector loop.
 */
class NioReactor private constructor(private val eventHandler: IoEventHandler) : Runnable {

    private val service: ExecutorService = Executors.newSingleThreadScheduledExecutor()
    private lateinit var channel: ServerSocketConnection
    @Volatile
    private var running = false

    companion object {
        @Throws(IOException::class)
        fun configure(port: Int): NioReactor = configure(port, 1)

        @Throws(IOException::class)
        fun configure(port: Int, poolSize: Int): NioReactor {
            val reactor = NioReactor(IoEventHandler(Executors.newFixedThreadPool(poolSize)))
            val channel = ServerSocketChannel.open()
            val selector = Selector.open()
            channel.bind(InetSocketAddress(port))
            channel.configureBlocking(false)
            channel.register(selector, SelectionKey.OP_ACCEPT)
            reactor.channel = ServerSocketConnection(selector, channel)
            return reactor
        }

        @Throws(IOException::class)
        fun connect(address: String, port: Int): NioReactor {
            val reactor = NioReactor(MSEventHandler())
            val selector = Selector.open()
            val channel = SocketChannel.open()
            channel.configureBlocking(false)
            channel.socket().keepAlive = true
            channel.socket().tcpNoDelay = true
            channel.connect(InetSocketAddress(address, port))
            channel.register(selector, SelectionKey.OP_CONNECT)
            reactor.channel = ServerSocketConnection(selector, channel)
            return reactor
        }
    }

    fun start() {
        running = true
        service.execute(this)
    }

    override fun run() {
        Thread.currentThread().name = "NioReactor"
        while (running) {
            try {
                channel.selector.select()
            } catch (e: IOException) {
                e.printStackTrace()
            }
            val iterator = channel.selector.selectedKeys().iterator()
            while (iterator.hasNext()) {
                val key = iterator.next()
                iterator.remove()
                try {
                    if (!key.isValid || !key.channel().isOpen) {
                        key.cancel()
                        continue
                    }
                    when {
                        key.isConnectable -> eventHandler.connect(key)
                        key.isAcceptable -> eventHandler.accept(key, channel.selector)
                        key.isReadable -> eventHandler.read(key)
                        key.isWritable -> eventHandler.write(key)
                    }
                } catch (t: Throwable) {
                    eventHandler.disconnect(key, t)
                }
            }
        }
    }

    fun terminate() {
        running = false
    }
}
